File: android/app/src/main/java/com/porttennanttandoori/pos/data/local/TokenStore.kt
package com.porttennanttandoori.pos.data.local

import android.content.Context
import androidx.datastore.preferences.core.edit
import androidx.datastore.preferences.core.stringPreferencesKey
import androidx.datastore.preferences.preferencesDataStore
import java.time.Instant
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.flow.map

private val Context.dataStore by preferencesDataStore(name = "pos_session")

/** What the app needs to authenticate a paired device: its id/secret (for re-login when the
 * access token expires) plus the current access token and when it stops being valid. */
data class DeviceSession(
    val deviceId: String,
    val deviceSecret: String,
    val accessToken: String,
    val accessTokenExpiresAt: Instant,
    val restaurantName: String,
) {
    val isAccessTokenExpired: Boolean
        get() = Instant.now().isAfter(accessTokenExpiresAt.minusSeconds(30))
}

/** Persists device pairing + the current access token across restarts. The secret never leaves
 * the device after pairing - it's only ever sent over HTTPS to device/login. */
class TokenStore(private val context: Context) {

    private object Keys {
        val DEVICE_ID = stringPreferencesKey("device_id")
        val DEVICE_SECRET = stringPreferencesKey("device_secret")
        val ACCESS_TOKEN = stringPreferencesKey("access_token")
        val ACCESS_TOKEN_EXPIRES_AT = stringPreferencesKey("access_token_expires_at")
        val RESTAURANT_NAME = stringPreferencesKey("restaurant_name")
    }

    val session: Flow<DeviceSession?> = context.dataStore.data.map { prefs ->
        val deviceId = prefs[Keys.DEVICE_ID] ?: return@map null
        val deviceSecret = prefs[Keys.DEVICE_SECRET] ?: return@map null
        val accessToken = prefs[Keys.ACCESS_TOKEN] ?: return@map null
        val expiresAt = prefs[Keys.ACCESS_TOKEN_EXPIRES_AT]?.let { Instant.parse(it) } ?: return@map null
        val restaurantName = prefs[Keys.RESTAURANT_NAME] ?: return@map null
        DeviceSession(deviceId, deviceSecret, accessToken, expiresAt, restaurantName)
    }

    suspend fun currentSession(): DeviceSession? = session.first()

    suspend fun savePairing(deviceId: String, deviceSecret: String) {
        context.dataStore.edit { prefs ->
            prefs[Keys.DEVICE_ID] = deviceId
            prefs[Keys.DEVICE_SECRET] = deviceSecret
        }
    }

    suspend fun saveAccessToken(accessToken: String, expiresAt: Instant, restaurantName: String) {
        context.dataStore.edit { prefs ->
            prefs[Keys.ACCESS_TOKEN] = accessToken
            prefs[Keys.ACCESS_TOKEN_EXPIRES_AT] = expiresAt.toString()
            prefs[Keys.RESTAURANT_NAME] = restaurantName
        }
    }

    /** Forgets everything, including the paired device secret - used when the device is
     * deactivated from the admin dashboard and 401s can no longer be recovered from. */
    suspend fun clear() {
        context.dataStore.edit { it.clear() }
    }
}

File: android/lib/api/token_store.dart
import 'package:flutter_secure_storage/flutter_secure_storage.dart';

/// What the app needs to authenticate a paired device: its id/secret (for
/// re-login when the access token expires) plus the current access token and
/// when it stops being valid. Mirrors TokenStore.kt / DeviceSession.kt.
class DeviceSession {
  final String deviceId;
  final String deviceSecret;
  final String accessToken;
  final DateTime accessTokenExpiresAt;
  final String restaurantName;

  DeviceSession({
    required this.deviceId,
    required this.deviceSecret,
    required this.accessToken,
    required this.accessTokenExpiresAt,
    required this.restaurantName,
  });

  bool get isAccessTokenExpired =>
      DateTime.now().isAfter(accessTokenExpiresAt.subtract(const Duration(seconds: 30)));
}

/// Persists device pairing + the current access token across restarts using
/// the platform keystore/keychain (flutter_secure_storage) - the secret never
/// leaves the device after pairing except over HTTPS to /api/auth/device/login.
class TokenStore {
  TokenStore({FlutterSecureStorage? storage}) : _storage = storage ?? const FlutterSecureStorage();

  final FlutterSecureStorage _storage;

  static const _kDeviceId = 'device_id';
  static const _kDeviceSecret = 'device_secret';
  static const _kAccessToken = 'access_token';
  static const _kAccessTokenExpiresAt = 'access_token_expires_at';
  static const _kRestaurantName = 'restaurant_name';

  Future<DeviceSession?> currentSession() async {
    final all = await _storage.readAll();
    final deviceId = all[_kDeviceId];
    final deviceSecret = all[_kDeviceSecret];
    final accessToken = all[_kAccessToken];
    final expiresAtRaw = all[_kAccessTokenExpiresAt];
    final restaurantName = all[_kRestaurantName];
    if (deviceId == null || deviceSecret == null || accessToken == null || expiresAtRaw == null || restaurantName == null) {
      return null;
    }
    return DeviceSession(
      deviceId: deviceId,
      deviceSecret: deviceSecret,
      accessToken: accessToken,
      accessTokenExpiresAt: DateTime.parse(expiresAtRaw),
      restaurantName: restaurantName,
    );
  }

  Future<void> savePairing(String deviceId, String deviceSecret) async {
    await _storage.write(key: _kDeviceId, value: deviceId);
    await _storage.write(key: _kDeviceSecret, value: deviceSecret);
  }

  Future<void> saveAccessToken(String accessToken, DateTime expiresAt, String restaurantName) async {
    await _storage.write(key: _kAccessToken, value: accessToken);
    await _storage.write(key: _kAccessTokenExpiresAt, value: expiresAt.toIso8601String());
    await _storage.write(key: _kRestaurantName, value: restaurantName);
  }

  /// Forgets everything, including the paired device secret - used when
  /// signing the terminal out.
  Future<void> clear() => _storage.deleteAll();
}
